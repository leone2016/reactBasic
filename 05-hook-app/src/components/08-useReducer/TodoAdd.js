import React from 'react'
import { useForm } from '../../hooks/useForm'
// clase 156
export const TodoAdd = ({ handleAddTodo }) => {
    const [{ description }, handleInputChange, reset] = useForm({
        description: ''
    });
    const handleSubmit = (e) => {
        e.preventDefault();

        if (description.trim().length <= 1) {
            return;
        }
        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false
        }
        handleAddTodo(newTodo)
        reset();
    }


    return (
        <form onSubmit={handleSubmit} >
            <input
                type="text"
                name="description"
                className="from-control"
                placeholder="Aprender ..."
                autoComplete="off"
                value={description}
                onChange={handleInputChange}
            />
            <button type="submit " className="btn btn-info" >Agregar TODO</button>
        </form>
    )
}
