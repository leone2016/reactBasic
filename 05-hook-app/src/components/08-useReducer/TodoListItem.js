import React from 'react'

export const TodoListItem = ({ todo,index, handleDelete, handleToggle }) => {
    return (
        <li key={todo.id} className="">
            <span onClick={() => handleToggle(todo.id)}
                className={`${todo.done && 'complete'}`} >
                {index + 1}) {todo.desc}
            </span>
            <button class="btn btn-danger"
                onClick={() => handleDelete(todo.id)}>
                Borrar
            </button>
        </li>

    )
}
