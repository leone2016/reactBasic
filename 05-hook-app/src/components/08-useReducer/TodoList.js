import React from 'react'
import { TodoListItem } from './TodoListItem'

export const TodoList = ({ todos, handleDelete, handleToggle }) => {
    return (
        <ul>
            {
                todos.map((todo, index) => (
                    <TodoListItem
                        key={todo.id}
                        todo={todo}
                        index={index}
                        handleDelete={handleDelete}
                        handleToggle={handleToggle}
                    />
                ))
            }
        </ul>
    )
}
