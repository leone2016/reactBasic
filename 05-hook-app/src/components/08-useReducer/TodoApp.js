import React, { useReducer, useEffect, useCallback } from 'react'
import { todoReducer } from './todoReducer'
import { TodoList } from './TodoList';
import { TodoAdd } from './TodoAdd';

const init = () => {
    return JSON.parse(localStorage.getItem('todo')) || [];
    /* return [{
        id: new Date().getTime(),
        desc: 'Aprender react',
        done: false
    
    }] */
}
export const TodoApp = () => {


    /**
     * recibe 3 parametros
     * useReducer: funcion reducer
     * initialState: estado inicial
     * init: funcion que se utilizar para inicializar el State en caso de que el state 
     *      sea procesado o haga varias acciones, cada que el componente se vuelve a redibujar
     * Ã² ejecutar esa accion, entonces al ser una funcion externa al componente lo hace mas ligero y rapido
     * incluso puede memorizar ese resultado en caso de necesitarlo
     */
    /*  const [state, dispatch] = useReducer(useReducer, initialState, init) */
    const [todos, dispatch] = useReducer(todoReducer, [], init);
 

    useEffect(() => {
        localStorage.setItem('todos', JSON.stringify(todos));
    }, [todos]);

   
    const handleAddTodo = (newTodo) => {
        dispatch({
            type: 'add',
            payload: newTodo
        });
    }
    const handleDelete = useCallback(
        (todoId) => {
            const action = {
                type: 'delete',
                payload: todoId
            }
            dispatch(action);
        },
        [],
    )

    const handleToggle = useCallback(
        (todoId) => {
            dispatch({
                type: 'toggle',
                payload: todoId
            });
        },
        [],
    )

    return (
        <div>
            <h2>TodoApp ({todos.length})</h2>

            <TodoAdd handleAddTodo={handleAddTodo}/>
            <hr />
            <TodoList
                todos={todos}
                handleDelete={handleDelete}
                handleToggle={handleToggle}
            />

        </div>
    )
}
