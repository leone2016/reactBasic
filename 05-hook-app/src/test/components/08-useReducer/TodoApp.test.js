import React from 'react';
import { shallow, mount } from 'enzyme';
import { TodoApp } from '../../../components/08-useReducer/TodoApp';
import { act } from '@testing-library/react';
import { demoTodos } from '../../fixtures/demoTodos';
// cada que se ocupe el componente en este caso TodoApp, es necesario import React form 'react';
//clase 157
describe('test suit <TodoApp/>', () => {
    const wrapper = shallow(<TodoApp />);
    // en la linea 32 de TodoApp se realiza un guardado en localStorage, 
    // esta es una forma generica de mockear la accion de guardar en localStorage
    Storage.prototype.setItem = jest.fn(()=>{});

    test('debe de mostrar el componente correctamente', () => {
        expect(wrapper).toMatchSnapshot();


    });

    test('debe de agregar un TODO', () => {
        // el mount se puede utilizar cuando se necesita probar toda la aplicacion en contexto
        // shalow vs mount
        /**
         * la diferencia esta en el nivel en el cual es renderizada la aplicacion, es que el shallow es mas basico 
         */
        const wrapper = mount(<TodoApp />);
        // este act si sale de la libreria de pruebas de react
        // se utiliza para hacer modificaciones
        act(() => {
            wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodos[0]);
            wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodos[1]);

        });

        expect(wrapper.find('h2').text().trim()).toBe('TodoApp (2)');
        expect( localStorage.setItem).toHaveBeenCalledTimes(2); // estas dos veces estan en Ã±a linea 30 y 31 de esta prueba 
        /**
         * esta validacion es similar a la que se hizo en TodoAdd.test.js linea 48 
         * esta validacion esta incompleta
         * expect( localStorage.setItem).toHaveBeenCalledwith({});
         *  */ 
        
    });

    test('debe de borrar un TODO',()=>{
        wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodos[0]);
        expect(wrapper.find('h2').text().trim()).toBe('TodoApp (1)'); // prie
        wrapper.find('TodoList').prop('handleDelete')(demoTodos[0].id);
        expect(wrapper.find('h2').text().trim()).toBe('TodoApp (0)');
    });
})