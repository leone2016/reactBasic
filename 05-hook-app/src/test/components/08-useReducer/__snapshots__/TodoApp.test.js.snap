// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test suit <TodoApp/> debe de mostrar el componente correctamente 1`] = `
<div>
  <h2>
    TodoApp (
    0
    )
  </h2>
  <TodoAdd
    handleAddTodo={[Function]}
  />
  <hr />
  <TodoList
    handleDelete={[Function]}
    handleToggle={[Function]}
    todos={Array []}
  />
</div>
`;
